internal/webhooks/vspherefailuredomain_webhook.go

```diff
@@ -0,0 +1,120 @@
+func (webhook *VSphereFailureDomainWebhook) ValidateCreate(_ context.Context, raw runtime.Object) (admission.Warnings, error) {
+	var allErrs field.ErrorList
+
+	obj, ok := raw.(*infrav1.VSphereFailureDomain)
+	if !ok {
+		return nil, apierrors.NewBadRequest(fmt.Sprintf("expected an VSphereFailureDomain but got a %T", raw))
+	}
+	if obj.Spec.Topology.ComputeCluster == nil && obj.Spec.Topology.Hosts != nil {
+		allErrs = append(allErrs, field.Forbidden(field.NewPath("spec", "Topology", "ComputeCluster"), "cannot be empty if Hosts is not empty"))
+	}
+
+	if obj.Spec.Region.Type == infrav1.HostGroupFailureDomain {
+		allErrs = append(allErrs, field.Forbidden(field.NewPath("spec", "Region", "Type"), fmt.Sprintf("region's Failure Domain type cannot be %s", obj.Spec.Region.Type)))
+	}
+
+	if obj.Spec.Zone.Type == infrav1.HostGroupFailureDomain && obj.Spec.Topology.Hosts == nil {
+		allErrs = append(allErrs, field.Forbidden(field.NewPath("spec", "Topology", "Hosts"), fmt.Sprintf("cannot be nil if zone's Failure Domain type is %s", obj.Spec.Zone.Type)))
+	}
+
+	if obj.Spec.Region.Type == infrav1.ComputeClusterFailureDomain && obj.Spec.Topology.ComputeCluster == nil {
+		allErrs = append(allErrs, field.Forbidden(field.NewPath("spec", "Topology", "ComputeCluster"), fmt.Sprintf("cannot be nil if region's Failure Domain type is %s", obj.Spec.Region.Type)))
+	}
+
+	if obj.Spec.Zone.Type == infrav1.ComputeClusterFailureDomain && obj.Spec.Topology.ComputeCluster == nil {
+		allErrs = append(allErrs, field.Forbidden(field.NewPath("spec", "Topology", "ComputeCluster"), fmt.Sprintf("cannot be nil if zone's Failure Domain type is %s", obj.Spec.Zone.Type)))
+	}
+
+	return nil, aggregateObjErrors(obj.GroupVersionKind().GroupKind(), obj.Name, allErrs)
+}
+
+// ValidateUpdate implements webhook.Validator so a webhook will be registered for the type.
+func (webhook *VSphereFailureDomainWebhook) ValidateUpdate(_ context.Context, oldRaw runtime.Object, newRaw runtime.Object) (admission.Warnings, error) {
+	oldTyped, ok := oldRaw.(*infrav1.VSphereFailureDomain)
+	if !ok {
+		return nil, apierrors.NewBadRequest(fmt.Sprintf("expected an VSphereFailureDomain but got a %T", oldRaw))
+	}
+	newTyped, ok := newRaw.(*infrav1.VSphereFailureDomain)
+	if !ok {
+		return nil, apierrors.NewBadRequest(fmt.Sprintf("expected an VSphereFailureDomain but got a %T", newRaw))
+	}
+	if !reflect.DeepEqual(newTyped.Spec, oldTyped.Spec) {
+		return nil, field.Forbidden(field.NewPath("spec"), "VSphereFailureDomainSpec is immutable")
+	}
+	return nil, nil
+}
+
+// ValidateDelete implements webhook.Validator so a webhook will be registered for the type.
+func (webhook *VSphereFailureDomainWebhook) ValidateDelete(_ context.Context, _ runtime.Object) (admission.Warnings, error) {
+	return nil, nil
+}
+
+// Default implements webhook.Defaulter so a webhook will be registered for the type.
+//
+//nolint:staticcheck
```

Comment:
Let's add the nolint directly at the lines where we have the findings. This ensures we don't accidentally cover other findings, and it's more clear that we're ignoring the deprecation finding

(I know the func is not that long at the moment, but it's more future-proof to ignore findings more precisely)

Reply:
Fixed.